### DISCUSSION TITLE
The Great Prefabification

### DISCUSSION THREAD

**Discussion by andybak** [icosa-gallery/open-brush#123] [Ideas and Feature Requests]
Merge conflicts in Unity assets are a pain. You can mitigate the pain but you can't make it go away.

It's a good idea to reduce the chance and scope of a merge conflict and one way to do this is to make extensive use of prefabs.

If your scene mainly consists of prefabs then the chance of a conflict is reduced to the situation where both branches have modified the same prefab.

I propose we do a sweep through the main scene and create prefabs of as many components as makes sense.

There are broadly speaking two opposing ways to interpret "makes sense": a maximalist and a minimalist interpretation.

Maximalist: Make prefabs out of almost everything. Including tested prefabs

Minimalist: Identify which components are likely to be modified and make prefabs out of just those.

Obviously these are the extemes and there are compromises between the two approaches.

Here are some things that are potentially likely to be modified in my experience to date:

1. App > Config (various settings - especially "experimental" and "SDK mode"
2. App > HttpServer (Well - modified this for the API feature but that might be an unusual case)
3. SketchControls > SketchSurface (mainly to add child tools)
4. SketchControls (Panel Map whenever a panel is added)

More broadly - most of the children of App and SketchControls might be good candidates. I've not had to touch the other root objects in the scene but that might be a factor of the kind of things I've been working on.

Thoughts?

**Comment by andybak** [comment]
The biggest downside to lots of prefabs is forgetting to apply overrides when you tweak the containing scene or prefab. but that's easily fixed and easily spotted when we accept a PR

**Comment by TimAidley** [comment]
I am in favour of doing this. I actually started doing some of this kind of work on the sketchbook panel and more recently the brush panel.
